Add: Configuration file functionality

oleorhagen opened the pull request:
This adds the possibility to add the username and password to a configuration
file, in which the 'mender-cli' tool will look if no password or username is set
on the CLI. The configuration file is expected to be JSON.

This helps usage, in that now, in order to login, a user with a configuration
file can do:

```console
$ mender-cli login
```

as opposed to:
```console
$ mender-cli --username foo --password bar --server bar.com
```

Changelog: Commit

Signed-off-by: Ole Petter <ole.orhagen@northern.tech>

oleorhagen commented:
Small value add addition.
This has only been tested on the CLI manually. Do you think automatic testing is required for this?

kacf commented on cmd/login.go:
```diff
@@ -62,24 +72,27 @@ type LoginCmd struct {
+	viper.BindPFlag(argRootServer, cmd.Flags().Lookup(argRootServer))
+	viper.BindPFlag(argLoginUsername, cmd.Flags().Lookup(argLoginUsername))
+	viper.BindPFlag(argLoginPassword, cmd.Flags().Lookup(argLoginPassword))
+	server := viper.GetString(argRootServer)
```
The documentation for viper is a bit confusing, so I'm just making sure I got it right: This *falls back* to the config file and command line arguments take precedence, right?

oleorhagen replied:
Yes, exactly :)

I also had to do a manual test, to figure out which was the default here

kacf reviewed (changes requested):
> This has only been tested on the CLI manually. Do you think automatic testing is required for this?

I think we should have it. :-) It should not be too hard though, if you test the `"."` location?

lluiscampos commented on coverage.txt:
```diff
@@ -0,0 +1,2 @@
+mode: set
+github.com/mendersoftware/mender-cli/main.go:20.13,22.2 1 0
```
Committed by mistake? Please remove.

oleorhagen replied:
thx

lluiscampos commented on LIC_FILES_CHKSUM.sha256:
```diff
@@ -3,20 +3,27 @@
 f42d670262f9aed37a33c97d93fa2c1324f439ee03d7690710c5fe002561b0b5  vendor/github.com/inconshreveable/mousetrap/LICENSE
 5e3400b93bbb099e83e52bab885e7441750673c21f97988ca3f1240639b63283  vendor/github.com/spf13/cobra/LICENSE.txt
-b8514c577c1c4b46cee454d5a882b15fa411e72c5bd7f801f241591789fce61a  vendor/github.com/spf13/pflag/LICENSE
+8514c577c1c4b46cee454d5a882b15fa411e72c5bd7f801f241591789fce61a  vendor/github.com/spf13/pflag/LICENSE
+bf6918bd93ffe07f4b2c61b8287c32cb3122e08aed0be50f1c7d0eddc87877a8f  vendor/gopkg.in/ini.v1/LICENSE
```
Smells like a mistake :)

lluiscampos replied:
I wonder how can the unit tests pass...

oleorhagen replied:
That is a little worrying, indeed. Looking into it

oleorhagen replied:
So, @lluiscampos I've dug into this a bit. And the error is that the `shasum` program does not error on this, as it is considered a warning only:
```
shasum: WARNING: 6 lines are improperly formatted
```

These warnings though, stem from the empty lines in the `LIC_FILES_CHECKSUM.sha256` file, which are also considered improperly formatted. 

I will add a fix to `mendertesting next.

lluiscampos replied:
@oleorhagen Awesome. Thank you for following up

oleorhagen replied:
Follow up PR: https://github.com/mendersoftware/mendertesting/pull/95/files

lluiscampos reviewed (changes requested):
Really nice feature :rocket: 

See two comments below.

lluiscampos commented on tests/tests/test_login.py:
```diff
@@ -19,78 +19,186 @@
+
+        r = c.run("login", "--skip-verify")
+
+        assert r.returncode != 0, r.stderr
```
I saw that in your last force-pushes you have removed the explicit check for the error message verification. Why?

oleorhagen replied:
Well, spotted!
They we're just copy pastes from the other tests. I removed them to see if they passed on their own.
Adding proper stderr expectations now.

lluiscampos replied:
> Well, spotted!

It is a new world since @olehermanse showed me the GitHub hidden feature of the "force-pushed" link. This has helped me greatly on reviews since. Only for that is worth to join CFEngine for an sprint! :)

lluiscampos reviewed (approved):
The tests look good. Well done :climbing:

oleorhagen commented:
Alright. Should be ready for primetime.

Only minor changes since last review:

* The tool notifies on the location of the configuration file used
* Readme is extended a little
* Test output assertions are now proper
